### ISSUE TITLE
Data Preprocessing

### ISSUE THREAD

**Issue by 00015775** [00015775/uzslr-isolated-dynamic#5]
# Depends on completion of #3 (train/val/test stratified split 80/10/10)

## Data Preprocessing

I will write here the rough sketches of doing the data-preprocessing

**Comment by 00015775** [comment]
Below are links that might help to pinpoint the exact coordinates of `eyes, nose, lips` from the MediaPipe Face Mesh coordinates:

[link-1](https://www.kaggle.com/discussions/questions-and-answers/393052)


[link-2](https://github.com/tensorflow/tfjs-models/blob/838611c02f51159afdd77469ce67f0e26b7bbb23/face-landmarks-detection/src/mediapipe-facemesh/keypoints.ts)

[link-3](https://github.com/google-ai-edge/mediapipe/blob/e0eef9791ebb84825197b49e09132d3643564ee2/mediapipe/modules/face_geometry/data/canonical_face_model_uv_visualization.png)

**Comment by 00015775** [comment]
Update (_NOT FINAL, UNDER REVIEW, NEEDS FURTHER TESTING!_)
-

<pre>
frame-XX.npy        -> (1662,)
32 stacked frames   -> (32, 1662)
Preprocess          -> returns (32, 708)
Dataset[i]          -> one repetition (x:sequence, y:label); x:(32,708) y:class index (0-49)
DataLoader batch    -> (B, 32, 708)
Model               -> predicts 1 sign label, based on (32, 708) input
</pre>

---

**For training split only:**
-
#### **Dataset Class:**
- 50 signs
- 40–41 repetitions per sign
    
Total samples ≈ 2000 repetitions  
Each `__getitem__` should return:  
**x**: Tensor of shape `(32, 708)`  
**y**: class index `(0–49) int`    
<pre>
    Dataset length ≈ 2000
    Dataset[i] = (sequence, label)
 </pre>

#### **DataLoader Class:**
- for instance **IF** `batch_size = 16`  
- DataLoader stacks each sample(repetition) along a new batch dimension
<pre>
(batch_size, T, features)
= (16, 32, 708)
</pre>
- This is what the model will be feed with `(B, 32, 708)`.
```python
def forward(self, x):
    # x.shape == (B, T, F)
    # (batch, time, features)
```

#### **Overall picture:**
<pre>
Filesystem
 └── rep-XX/
     └── 32 frames (.npy)

Dataset[i]
 ├── loads 32 frames
 ├── x: (32, 708)
 └── y: label 0-49 (int)

DataLoader
 ├── batch of repetitions
 └── x: (B, 32, 708)

Model
 └── predicts ONE label per repetition, based on (32, 708) input
</pre>

**Comment by 00015775** [comment]
UPDATE:
-
Everything is working as expected, update `preprocessing/README.md` documentation where it talks about Data Preprocessing and Augmentation pipeline, leave other details as they are. Then create a new `issue` and write out the requirements for model design and evaluation metrics, create a new `./training` folder with `notebooks`, `scripts` and other required subfolders and make sure to keep track of hyperparameters and save them in order to compare and decide which values works the best. 

> [!IMPORTANT]
> Preprocessing is done on `CPU`
> Model training is done on `GPU`

**Comment by 00015775** [comment]
Also UPDATE the main `README.md` and provide relevant links and gifs
